pkg/providers/cloudstack/cloudstack.go

```diff
@@ -551,10 +551,23 @@ func (p *cloudstackProvider) needsNewKubeadmConfigTemplate(workerNodeGroupConfig
 	return true, nil
 }
 
-// AnyImmutableFieldChanged Used by EKS-A controller and CLI upgrade workflow to compare generated CSDC/CSMC's from
+// NeedNewMachineTemplate Used by EKS-A controller and CLI upgrade workflow to compare generated CSDC/CSMC's from
 // CAPC resources in fetcher.go with those already on the cluster when deciding whether or not to generate and apply
 // new CloudStackMachineTemplates.
-func AnyImmutableFieldChanged(generatedCsdc, actualCsdc *v1alpha1.CloudStackDatacenterConfig, generatedCsmc, actualCsmc *v1alpha1.CloudStackMachineConfig, log logr.Logger) bool {
+func NeedNewMachineTemplate(generatedCsdc, actualCsdc *v1alpha1.CloudStackDatacenterConfig, generatedCsmc, actualCsmc *v1alpha1.CloudStackMachineConfig, log logr.Logger) bool {
+	if notMatch := compareDatacenterConfig(generatedCsdc, actualCsdc, log); notMatch {
+		return true
+	}
+
+	if err := v1alpha1.ValidateImmutableFieldsCloudStackMachineConfig(generatedCsmc, actualCsmc); len(err) > 0 {
```

Comment:
Based on the name alone, I would expect this to never return an error because none of the fields it checks should have changed and that should probably be validated prior to executing this logic. 

What do you think?